wasm2c/wasm-rt-impl.c

```diff
@@ -154,35 +344,108 @@ uint32_t wasm_rt_grow_memory(wasm_rt_memory_t* memory, uint32_t delta) {
 #endif
 #endif
+
 #if WABT_BIG_ENDIAN
-  memmove(new_data + new_size - old_size, new_data, old_size);
-  memset(new_data, 0, delta_size);
+  memmove(memory->data + new_size - old_size, memory->data, old_size);
+  memset(memory->data, 0, delta_size);
 #endif
   memory->pages = new_pages;
   memory->size = new_size;
-  memory->data = new_data;
+#if defined(WASM_CHECK_SHADOW_MEMORY)
+  wasm2c_shadow_memory_expand(memory);
+#endif
   return old_pages;
 }
 
 void wasm_rt_allocate_table(wasm_rt_table_t* table,
                             uint32_t elements,
                             uint32_t max_elements) {
+  assert(max_elements >= elements);
   table->size = elements;
   table->max_size = max_elements;
   table->data = calloc(table->size, sizeof(wasm_rt_elem_t));
+  assert(table->data != 0);
+}
+
+void wasm_rt_deallocate_table(wasm_rt_table_t* table) {
+  free(table->data);
 }
+
+#define WASM_SATURATING_U32_ADD(ret_ptr, a, b) \
+  {                                            \
+    if ((a) > (UINT32_MAX - (b))) {            \
+      /* add will overflowed */                \
+      *ret_ptr = UINT32_MAX;                   \
+    } else {                                   \
+      *ret_ptr = (a) + (b);                    \
+    }                                          \
+  }
+
+#define WASM_CHECKED_U32_RET_SIZE_T_MULTIPLY(ret_ptr, a, b)     \
+  {                                                             \
+    if ((a) > (SIZE_MAX / (b))) {                               \
+      /* multiple will overflowed */                            \
+      wasm_rt_trap(WASM_RT_TRAP_CALL_INDIRECT_TABLE_EXPANSION); \
+    } else {                                                    \
+      /* convert to size by assigning */                        \
+      *ret_ptr = a;                                             \
+      *ret_ptr = *ret_ptr * b;                                  \
+    }                                                           \
+  }
+
+void wasm_rt_expand_table(wasm_rt_table_t* table) {
+  uint32_t new_size = 0;
+  WASM_SATURATING_U32_ADD(&new_size, table->size, 32);
+
+  if (new_size > table->max_size) {
+    new_size = table->max_size;
+  }
+
+  if (table->size == new_size) {
+    // table is already as large as we allowed, can't expand further
+    wasm_rt_trap(WASM_RT_TRAP_CALL_INDIRECT_TABLE_EXPANSION);
+  }
+
+  size_t allocation_size = 0;
+  WASM_CHECKED_U32_RET_SIZE_T_MULTIPLY(&allocation_size, sizeof(wasm_rt_elem_t),
+                                       new_size);
+  table->data = realloc(table->data, allocation_size);
+  assert(table->data != 0);
+
+  memset(&(table->data[table->size]), 0,
+         allocation_size - (table->size * sizeof(wasm_rt_elem_t)));
+  table->size = new_size;
+}
+
+#undef WASM_PAGE_SIZE
+#undef WASM_HEAP_GUARD_PAGE_SIZE
+#undef WASM_HEAP_ALIGNMENT
+#undef WASM_HEAP_DEFAULT_MAX_PAGES
+#undef WASM_HEAP_MAX_ALLOWED_PAGES
+#undef WASM_SATURATING_U32_ADD
+#undef WASM_CHECKED_U32_RET_SIZE_T_MULTIPLY
```

Comment:
One of the things Firefox does is to unify source files during build to improve compilation performance. I have to double check, but I think all of the runtime files in the `wasm2c` maybe be built as a unified source file

Reply:
Thats interesting.    Seems reasonable I suppose, but if we want/need to support that mode we would need to add some kind of testing to prevent regressions.